chain/chain/src/chain.rs

```diff
@@ -1767,26 +1771,79 @@ impl Chain {
-            let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
-            let part = self.store.owned_store().get(ColStateParts, &key)?.unwrap();
+        if let Some(result) = prev_execution_result {
+            if result.shard_id != shard_id || result.epoch_id != epoch_id {
+                // Results are not for this shard
+                return Ok(SetStateFinalizeResult::NoAction);
+            }
+            match current_task {
+                Some(scheduled_part_id) => {
+                    if *scheduled_part_id != result.part_id {
+                        // We had probably had to abort process last time
+                        return Ok(SetStateFinalizeResult::NoAction);
+                    }
+                }
+                None => {
+                    // We had probably had to abort process last time
+                    return Ok(SetStateFinalizeResult::NoAction);
+                }
+            }
+            if result.part_id + 1 < num_parts {
+                let part_id = result.part_id + 1;
+                let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
+                let part = self.store.owned_store().get(ColStateParts, &key)?.unwrap();
 
-            self.runtime_adapter.apply_state_part(
-                shard_id,
-                &state_root,
-                part_id,
-                num_parts,
-                &part,
-                &epoch_id,
-            )?;
+                state_parts_task_scheduler(StatePartsMessage {
+                    shard_id,
+                    state_root,
+                    part_id,
+                    num_parts,
+                    part,
+                    epoch_id,
+                    source,
+                });
+                self.currently_applying_state_part.insert(map_key, part_id);
+                return Ok(SetStateFinalizeResult::ScheduledNext);
+            } else {
+                self.currently_applying_state_part.remove(&map_key);
+            }
+        } else {
+            if current_task.is_some() {
+                // We are already doing something for this epoch and shard
+                return Ok(SetStateFinalizeResult::NoAction);
+            }
+            if num_parts != 0 {
+                let part_id = 0;
+                let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
+                let part = self.store.owned_store().get(ColStateParts, &key)?.unwrap();
+
+                state_parts_task_scheduler(StatePartsMessage {
+                    shard_id,
+                    state_root,
+                    part_id,
+                    num_parts,
+                    part,
+                    epoch_id,
+                    source,
+                });
+                self.currently_applying_state_part.insert(map_key, part_id);
+                return Ok(SetStateFinalizeResult::ScheduledNext);
+            }
         }
 
+        // We've done with applying parts
+
```

Comment:
This probably wants

`assert!(num_parts == 0 && self.currently_applying_state_part.is_empty())`

assert (not sure if the condition I've got is right).